text/0000-public-private-dependencies.md

```diff
@@ -0,0 +1,199 @@
+- Feature Name: `public_private_dependencies`
+- Start Date: 2017-04-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Introduce a public/private distinction to crate dependencies.
+
+# Motivation
+[motivation]: #motivation
+
+The crates ecosystem has greatly expanded since Rust 1.0 and with that a few patterns for
+dependencies have evolved that challenge the currently existing dependency declaration
+system in cargo and rust.  The most common problem is that a crate `A` depends on
+another crate `B` but some of the types from crate `B` are exposed through the API in
+crate `A`.  This causes problems in practice if that dependency `B` is also used by the
+user's code itself which often leaves users in less than ideal situations where either
+their code refuses to compile because different versions of those libraries are requested
+or where compiler messages are less than clear.
+
+The introduction of an explicit distinction between public and private dependencies can
+solve some of these issues and also let us lift some restrictions that should make some
+code compile that previously was prevented from compiling by restrictions in cargo.
+
+**Q: What is a public dependency?**<br>
+A: a dependency is public if some of the types or trait of that dependency is itself
+exported through the main crate.  The most common places where this happens is obviously
+return values and function parameter but obviously the same applies to trait implementations
+and many other things.  Because public can be tricky to determine for a user this RFC
+proposes to extend the compiler infrastructure to detect the concept of "public dependency".
+This will help the user understanding this concept and avoid making mistakes in
+the `Cargo.toml`
+
+Effectively the idea is that if your own library bumps a public dependency it means that
```

Comment:
Wouldn't this only be true for bumping the major version of a public dependency?

For example, say crate `A` version `1.0.0` depends on crate `B` version `2.0.0`. Can `A` upgrade to `B` version `2.0.1` in a patch release (e.g. `A` version `1.0.1`)?